### PR TITLE
feat(streaming): enable dml executor to pause and resume on scaling

### LINKED ISSUE #8056: bug(streaming): DMLExecutor should handle Mutation::Pause and Mutation::Resume
DMLExecutor should handle Mutation::Pause and Mutation::Resume. When a Mutation::Pause barrier comes, the DMLExecutor should stop pulling the operations from DMLHandle channel util the Mutation::Resume comes. It needs to make sure that there are no stream chunks between Mutation::Pause and Mutation::Resume.

### PR DESCRIPTION
**Pull request by xx01cyx**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Enable dml executor to pause and resume on scaling.
- A little refactor on `StreamReaderWithPause` (previously named `SourceReaderStream`):
   - Make the left arm accept general message types instead of barriers only.
   - Introduce non-biased `StreamReaderWithPause`.

Fixes #8056 

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/stream/src/executor/stream_reader.rs [hunk 1]
```diff
@@ -25,32 +25,35 @@ use risingwave_connector::source::{BoxSourceWithStateStream, StreamChunkWithStat
...
+use crate::executor::{Barrier, Message};
+
+type ExecutorMessageStream = BoxStream<'static, StreamExecutorResult<Message>>;
+type ReaderStreamData = StreamExecutorResult<Either<Message, StreamChunkWithState>>;
+type ReaderArm = BoxStream<'static, ReaderStreamData>;
+type ReaderStreamWithPauseInner =
+    SelectWithStrategy<ReaderArm, ReaderArm, impl FnMut(&mut PollNext) -> PollNext, PollNext>;
+
+/// `ReaderStreamWithPause` merges two streams, with one of them receiving barriers. When the
+/// barrier requires the data stream
```
Comment (BugenZhao) [inline src/stream/src/executor/stream_reader.rs:37]:
```suggestion
/// barrier requires the data stream, ...
```

Comment (BugenZhao) [reply src/stream/src/executor/stream_reader.rs:37]: What about mentioning which side will be biased when `BIASED` is `true` somewhere?

File: src/stream/src/executor/stream_reader.rs [hunk 2]
```diff
@@ -64,38 +67,50 @@ impl SourceReaderStream {
...
+    }
+
+    /// Construct a `ReaderStreamWithPause` with one stream receiving barriers only and the other
+    /// receiving data only.
+    pub fn new_with_barrier_receiver(
         barrier_receiver: UnboundedReceiver<Barrier>,
-        source_stream: BoxSourceWithStateStream,
+        data_stream: BoxSourceWithStateStream,
     ) -> Self {
+        let barrier_receiver_arm = Self::barrier_receiver(barrier_receiver)
```
Comment (BugenZhao) [inline src/stream/src/executor/stream_reader.rs:91]: Call we directly call `new_with_message_stream` with `Self::barrier_receiver(barrier_receiver)` here?

**Review by BugenZhao — COMMENTED** [review]
Really elegant implementation. Rest LGTM!

**Review by st1page — APPROVED** [review]
LGTM

**Review by waruto210 — APPROVED** [review]
LGTM.

**Pull request merged**